### PR TITLE
Optimize val::array() iterator version for numeric types.

### PR DESCRIPTION
**Pull request by xbcnn**
Again numeric contiguous array-like data can be optimized same as #17292.
But this needs c++20 and later.

Also updated the test cases and benchmark tests.

### CODE REVIEW

**Comment by xbcnn** [comment]
@RReverser @sbc100  Please take a look.  Thanks.

File: system/include/emscripten/val.h [hunk 1]
```diff
@@ -327,6 +327,17 @@ class val {
 
     template<typename Iter>
     static val array(Iter begin, Iter end) {
+#if _LIBCPP_STD_VER > 17
+        if constexpr (std::contiguous_iterator<Iter> &&
+                      internal::typeSupportsMemoryView<
+                        typename std::iterator_traits<Iter>::value_type>()) {
+            val view{typed_memory_view(end - begin, std::to_address(begin))};
```
[the remark this answers is not in the document]
Comment (xbcnn) [reply system/include/emscripten/val.h:334]: I think it's OK here as we first make sure they are contiguous iterators.  From `std::random_access_iterator_tag` above, `end-start` is the distance( or size).  Anyway I will use [`std::distance`](url: https://en.cppreference.com/w/cpp/iterator/distance) for safety.

[the remark this answers is not in the document]
Comment (RReverser) [reply system/include/emscripten/val.h]: > I think it's OK here as we first make sure they are contiguous iterators.

Yeah, but looking at the list of guarantees I'm not sure why they said specifically [std::iter_difference_t](http://en.cppreference.com/w/cpp/iterator/iter_t)<I>(b - a) and not just `b - a`, so I'd rather be safe.

image

[the remark this answers is not in the document]
Comment (RReverser) [reply system/include/emscripten/val.h]: > btw, seems there are typo errors in [`std::random_access_iterator`](https://en.cppreference.com/w/cpp/iterator/random_access_iterator):
> 
> std::addressof(a += n) is equal to std::addressof(a). ... std::addressof(b -= n) is equal to std::addressof(b).

Hmm yeah that seems... bad.

**Comment by xbcnn** [comment]
@RReverser @sbc100  Is there any further concerns about this PR?

File: system/include/emscripten/val.h [hunk 2]
```diff
@@ -327,6 +327,17 @@ class val {
 
     template<typename Iter>
     static val array(Iter begin, Iter end) {
+#if _LIBCPP_STD_VER > 17
```
[the remark this answers is not in the document]
Comment (xbcnn) [reply system/include/emscripten/val.h]: Done.  But I think they're same thing :)

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
lgtm.. @RReverser any more comments?

**Review by RReverser — APPROVED** [review]

**Pull request merged**